pkg/provider/terraform/instance/plugin.go

```diff
@@ -1005,6 +1013,7 @@ func (p *plugin) Destroy(instID instance.ID, context instance.Context) error {
 	// Acquire Lock outside of recursive doDestroy function
 	p.fsLock.Lock()
 	defer p.fsLock.Unlock()
+	defer p.clearCachedInstances()
 
```

Comment:
Is this really guarding the cache?  The clearing comes after the lock is released.   I'm  also not sure if multiple defers is stylistically...  how about 
```
defer func() {
    clearCache()
    lock.Unlock()
}()
```

Reply:
Another thing. What about errors?  The defer will *always* clear cache even if we encounter an error.  Is that a good thing?

Reply:
Heh https://medium.com/intuitionmachine/the-strange-loop-in-alphago-zeros-self-play-6e3274fcdd9f?source=linkShare-cc3cd841bbd8-1515523953

Forgot multiple defers are last in fist out (it's pushed onto a stack).  Anyway my point remains for readability:)

Reply:
@chungers the `defer` functions are pushed on a stack, so they are executed in reverse order: https://stackoverflow.com/questions/32541870/multiple-defers-vs-deferred-anonymous-function

But I do see the value is having a utility function for this, then it's consistent and the order is obviousl

Reply:
@chungers yeah, having it clear the cache on error probably isn't ideal. But, IMO, we'd want to error on the side or refreshing the cache vs attempting to differentiate between errors that invalidated the cache and those that didn't.

Reply:
1.  Yeah I prefer the closure approach for readability.  I wouldn't make a separate function just for that though since that utility function would look weird just releasing a lock.   

2. Ok that's reasonable.  I think it's a matter of policy and at this point I would emphasize correctness too.  So yes - let's clear the cache on all mutations.